File: factory/hooks/memory-save.py
#!/usr/bin/env python3
"""
Stop hook: reminds Droid to save learnings to memory before ending.

On first stop (stop_hook_active=false): blocks and asks Droid to save memory.
On second stop (stop_hook_active=true): allows Droid to stop normally.
"""
import json
import os
import subprocess
import sys
from pathlib import Path


def _resolveProjectName(cwd):
    """Return git root basename if available, otherwise cwd basename."""
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--show-toplevel"],
            cwd=cwd,
            capture_output=True,
            text=True,
            timeout=3,
        )
        if result.returncode == 0 and result.stdout.strip():
            return os.path.basename(result.stdout.strip())
    except (subprocess.TimeoutExpired, FileNotFoundError, OSError):
        pass
    return os.path.basename(cwd)


try:
    inputData = json.load(sys.stdin)
except json.JSONDecodeError:
    sys.exit(0)

# If already triggered once this turn, let Droid stop
stopHookActive = inputData.get("stop_hook_active", False)
if stopHookActive:
    sys.exit(0)

cwd = inputData.get("cwd", os.getcwd())

# Read locked project name from session start, fallback to resolving fresh
lockFile = Path.home() / ".cli-tweaks" / ".session-locks" / str(os.getppid())
try:
    projectName = lockFile.read_text(encoding="utf-8").strip()
except (FileNotFoundError, OSError):
    projectName = _resolveProjectName(cwd)

memoryDir = Path.home() / ".cli-tweaks" / "memory" / projectName

# Ensure memory directory exists
memoryDir.mkdir(parents=True, exist_ok=True)

# Check if memory file exists
memoryFile = memoryDir / "MEMORY.md"
hasMemory = memoryFile.exists()

# Read the memory file once to drive two checks: line count (offload prompt)
# and presence of the required '## CRITICAL RULES' section (format/migration).
# Threshold is below the 200-line cap so the agent trims BEFORE overflowing.
MEMORY_SOFT_LIMIT = 180
lineCount = 0
hasCriticalSection = False
if hasMemory:
    try:
        memoryLines = memoryFile.read_text(encoding="utf-8").splitlines()
        lineCount = len(memoryLines)
        hasCriticalSection = any(
            line.strip().lower() == "## critical rules" for line in memoryLines
        )
    except (OSError, IOError):
        lineCount = 0

TEMPLATE = (
    "Use this MEMORY.md structure (sections in this order):\n"
    "  ## CRITICAL RULES        - non-negotiable active rules, imperative mood\n"
    "  ## Architecture & Config Facts - stable technical context (not rules)\n"
    "  ## Active Warnings       - pitfalls and recurring mistakes\n"
    "  ## Topic Files           - pointers to detail files (e.g. history.md)\n"
    "Write each bullet on a single line; do NOT hard-wrap text mid-bullet. "
    "Hard-wrapping inflates the line count and weakens the 40-line fallback reminder.\n"
)

if hasMemory:
    reason = (
        "Before stopping: if you learned an ACTIVE RULE that changes future behavior "
        "(build/test commands, an architecture fact, a user preference, a workflow rule), "
        "update {dir}/MEMORY.md or a topic file. Write rules in imperative mood. "
        "Put durable behavior rules under the '## CRITICAL RULES' section. "
        "Do NOT save commit hashes, dated fix histories, or archival narrative — "
        "put any historical detail in history.md, not MEMORY.md. "
        "If nothing new was learned and the format is already correct, just stop. "
        "Keep MEMORY.md under 200 lines. IMPORTANT: Always write memory in English only."
    ).format(dir=memoryDir)
else:
    reason = (
        "Before stopping: this is a new project with no memory yet. "
        "Create {dir}/MEMORY.md following the template below. "
        "Write rules in imperative mood. Do NOT save commit hashes or dated history. "
        "Keep it concise (under 200 lines). IMPORTANT: Always write memory in English only. "
        "If this was a trivial session with nothing worth remembering, just stop.\n"
        + TEMPLATE
    ).format(dir=memoryDir)

if hasMemory and not hasCriticalSection:
    reason += (
        "\nMANDATORY MIGRATION: MEMORY.md is MISSING the '## CRITICAL RULES' section, "
        "so it is NOT in the required format. You MUST restructure the whole file this "
        "session into the four-section template, in this exact order:\n"
        "  ## CRITICAL RULES        - non-negotiable active rules, imperative mood\n"
        "  ## Architecture & Config Facts - stable technical context (not rules)\n"
        "  ## Active Warnings       - pitfalls and recurring mistakes\n"
        "  ## Topic Files           - pointers to detail files (e.g. history.md)\n"
        "Preserve all real content, reorganize it under those sections, and convert "
        "rules to imperative mood. Do this before stopping."
    )

if hasMemory and lineCount >= MEMORY_SOFT_LIMIT:
    reason += (
        "\nOFFLOAD: MEMORY.md is now {n} lines, near the 200-line cap. "
        "Move the OLDEST or least-critical entries (resolved warnings, superseded "
        "facts, dated notes) into a topic file (e.g. history.md), keeping MEMORY.md "
        "a lean index of ACTIVE rules and current architecture facts."
    ).format(n=lineCount)

output = {
    "decision": "block",
    "reason": reason,
}
print(json.dumps(output))
sys.exit(0)

File: opencode/plugins/memory-save.ts
import type { Plugin } from "@opencode-ai/plugin"
import { homedir } from "node:os"
import { join, basename } from "node:path"
import { existsSync, mkdirSync, writeFileSync, rmSync } from "node:fs"
import { execSync } from "node:child_process"

/**
 * Stop hook: reminds the agent to save learnings to memory before ending.
 *
 * Port of claude/hooks/memory-save.py. On the first stop attempt this blocks by
 * injecting a continuation prompt; on the follow-up stop (the result of our own
 * prompt) it lets the agent stop. A per-session toggle file replaces Claude's
 * `stop_hook_active` flag, which has no documented OpenCode equivalent.
 *
 * NOT runtime-tested: authored against the documented `stop` hook signature
 * (input.sessionID + client.session.prompt). OpenCode was not installed.
 */

/** Return the git root basename if available, otherwise the cwd basename. */
function resolveProjectName(cwd: string): string {
  try {
    const root = execSync("git rev-parse --show-toplevel", {
      cwd,
      encoding: "utf8",
      timeout: 3000,
      stdio: ["ignore", "pipe", "ignore"],
    }).trim()
    if (root) return basename(root)
  } catch {
    // Not a git repo or git unavailable; fall back to the cwd basename.
  }
  return basename(cwd)
}

const TEMPLATE =
  "Use this MEMORY.md structure (sections in this order):\n" +
  "  ## CRITICAL RULES        - non-negotiable active rules, imperative mood\n" +
  "  ## Architecture & Config Facts - stable technical context (not rules)\n" +
  "  ## Active Warnings       - pitfalls and recurring mistakes\n" +
  "  ## Topic Files           - pointers to detail files (e.g. history.md)\n" +
  "Write each bullet on a single line; do NOT hard-wrap text mid-bullet. " +
  "Hard-wrapping inflates the line count and weakens the 40-line fallback reminder.\n"

export const MemorySavePlugin: Plugin = async ({ directory, client }) => {
  return {
    stop: async (input: any) => {
      const sessionId = input?.sessionID || input?.session_id
      if (!sessionId) return

      // Per-session toggle replaces Claude's stop_hook_active flag.
      // Absent  -> first stop: block, and record that we fired.
      // Present -> our own prompt's stop: clear it and let the agent stop.
      const guardDir = join(homedir(), ".cli-tweaks", ".stop-fired")
      const guardFile = join(guardDir, sessionId)
      try {
        mkdirSync(guardDir, { recursive: true })
      } catch {
        // Fail safe: if the state dir is unwritable, do not block the stop.
        return
      }
      if (existsSync(guardFile)) {
        rmSync(guardFile, { force: true })
        return
      }

      const projectName = resolveProjectName(directory)
      const memoryDir = join(homedir(), ".cli-tweaks", "memory", projectName)
      mkdirSync(memoryDir, { recursive: true })
      const hasMemory = existsSync(join(memoryDir, "MEMORY.md"))

      const reason = hasMemory
        ? "Before stopping: if you learned an ACTIVE RULE that changes future behavior " +
          "(build/test commands, an architecture fact, a user preference, a workflow rule), " +
          `update ${memoryDir}/MEMORY.md or a topic file. Write rules in imperative mood. ` +
          "Put durable behavior rules under the '## CRITICAL RULES' section. " +
          "Do NOT save commit hashes, dated fix histories, or archival narrative — " +
          "put any historical detail in history.md, not MEMORY.md. " +
          "MIGRATION: if MEMORY.md has no '## CRITICAL RULES' section, restructure the " +
          "whole file into the template below this session (preserve all real content, " +
          "just reorganize and convert rules to imperative mood). " +
          "If nothing new was learned and the format is already correct, just stop. " +
          "Keep MEMORY.md under 200 lines. IMPORTANT: Always write memory in English only.\n" +
          TEMPLATE
        : "Before stopping: this is a new project with no memory yet. " +
          `Create ${memoryDir}/MEMORY.md following the template below. ` +
          "Write rules in imperative mood. Do NOT save commit hashes or dated history. " +
          "Keep it concise (under 200 lines). IMPORTANT: Always write memory in English only. " +
          "If this was a trivial session with nothing worth remembering, just stop.\n" +
          TEMPLATE

      // Write the toggle BEFORE prompting so the follow-up stop is recognized.
      writeFileSync(guardFile, "1")
      await client.session.prompt({
        path: { id: sessionId },
        body: { parts: [{ type: "text", text: reason }] },
      })
    },
  }
}
